File: apps/worker/app/events.py
"""Cross-process real-time event bus (worker side).

Mirrors ``apps/api/src/ws/events.ts``: publish a ``{workspaceId, event}`` envelope onto the
shared Redis ``ws:events`` channel. The Fastify API subscribes and fans each event out to
the workspace's connected WebSocket clients. Best-effort — a publish failure never breaks
the job that emitted it.
"""
import json
import logging
from typing import Any

log = logging.getLogger("worker.events")

WS_CHANNEL = "ws:events"


async def publish_event(redis: Any, workspace_id: str, event: dict[str, Any]) -> None:
    """Publish one real-time event to a workspace. Swallows its own errors."""
    try:
        payload = json.dumps({"workspaceId": workspace_id, "event": event})
        await redis.publish(WS_CHANNEL, payload)
    except Exception:  # never let real-time delivery break a job
        log.warning("failed to publish ws event for workspace %s", workspace_id, exc_info=True)

File: apps/api/src/ws/events.ts
import type { WebSocketEvent } from '@growthos/types'
import { getRedis } from '../jobs/client.js'

/** Redis pub/sub channel carrying every real-time event across processes (API + worker). */
export const WS_CHANNEL = 'ws:events'

/** Wire envelope on the channel: the target workspace plus the client-facing event. */
export interface WsEnvelope {
  workspaceId: string
  event: WebSocketEvent
}

/**
 * Publish a real-time event to a workspace. Fire-and-forget and best-effort: it swallows
 * its own errors (same contract as `recordAudit`) so a bus failure can never break the
 * mutation that emitted it. The API's WS server (and any other API process) receives it
 * over Redis and fans it out to that workspace's connected sockets.
 */
export async function publishEvent(workspaceId: string, event: WebSocketEvent): Promise<void> {
  try {
    const payload: WsEnvelope = { workspaceId, event }
    await getRedis().publish(WS_CHANNEL, JSON.stringify(payload))
  } catch {
    // Never throw — real-time delivery is additive; clients still poll.
  }
}
